File: Arbitration/MPArbitration/Model/BaseFee.cs
using System.Text.Json.Serialization;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;

namespace MPArbitration.Model
{
    public abstract class BaseFee
    {
        [JsonPropertyName("id")]
        public int Id { get; set; } = 0;

        [JsonPropertyName("createdBy")]
        [StringLength(60)]
        public string CreatedBy { get; set; } = "";

        [JsonPropertyName("createdOn")]
        public DateTime? CreatedOn { get; set; } = null;

        [JsonPropertyName("dueDaysAfterColumnName")]
        public int DueDaysAfterColumnName { get; set; } = 0;

        [JsonPropertyName("dueDayType")]
        [Column(TypeName = "nvarchar(20)")]
        public DeadlineType DueDayType { get; set; } = DeadlineType.CalendarDays;

        [JsonPropertyName("feeAmount")]
        public double FeeAmount { get; set; }

        [JsonPropertyName("feeName")]
        [StringLength(60)]
        public string FeeName { get; set; } = "";

        [JsonPropertyName("feeType")]
        [Column(TypeName = "nvarchar(60)")]
        public FeeType FeeType { get; set; } = FeeType.Administrative;

        [JsonPropertyName("isActive")]
        public bool IsActive { get; set; }

        [JsonPropertyName("isRefundable")]
        public bool IsRefundable { get; set; }

        [JsonPropertyName("isRequired")]
        public bool IsRequired { get; set; }

        // Enter the name of a date column present on the Dispute.
        // If it exists and has a value when the fee is instantiated on the Dispute,
        // the fee deadline will be initialized based on the value.
        [JsonPropertyName("referenceColumnName")]
        [StringLength(40)]
        public string ReferenceColumnName { get; set; } = "";

        [JsonPropertyName("sizeMax")]
        public int SizeMax { get; set; } = 0;

        [JsonPropertyName("sizeMin")]
        public int SizeMin { get; set; } = 0;


        [JsonPropertyName("updatedBy")]
        [StringLength(60)]
        public string UpdatedBy { get; set; } = "";

        [JsonPropertyName("updatedOn")]
        public DateTime? UpdatedOn { get; set; } = null;
    }
}

File: Arbitration/MPArbitration/ClientApp/src/app/model/base-fee.ts
import { DeadlineType } from "./deadline-type-enum";
import { FeeType } from "./fee-type-enum";

export class BaseFee {
    public id = 0;
    public createdBy = '';
    public createdOn: Date|undefined;
    public dueDaysAfterColumnName = 0;
    public dueDayType:DeadlineType = DeadlineType.CalendarDays;
    public feeAmount = 0;
    public feeName = '';
    public feeType:FeeType = FeeType.Administrative;
    public isActive = true;
    public isRefundable = false;
    public isRequired = false;
    public referenceColumnName = '';
    public sizeMin = 0;
    public sizeMax = 0;
    public updatedBy = '';
    public updatedOn: Date|undefined;

    constructor(obj?:any){
        if(!obj)
            return;

        Object.assign(this, JSON.parse(JSON.stringify(obj)));

        if(typeof obj.dueDayType === 'string') {
            const t = obj.dueDayType as keyof typeof DeadlineType;
            this.dueDayType = DeadlineType[t] ?? DeadlineType.CalendarDays;
        }
        
        if(obj.dueDayType === undefined)
            this.dueDayType = DeadlineType.CalendarDays; // this is the safest default

        if(typeof obj.feeType === 'string') {
            const t = obj.feeType as keyof typeof FeeType;
            this.feeType = FeeType[t] ?? FeeType.Administrative;
        }
        
        if(obj.feeType === undefined)
            this.feeType = FeeType.Administrative; // Administrative won't trigger any calculations
    }
}
